volatility3/framework/plugins/linux/lsblk.py

```diff
@@ -0,0 +1,388 @@
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        return [
+            requirements.ModuleRequirement(
+                name="kernel",
+                description="Linux kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+            requirements.VersionRequirement(
+                name="linuxutils", component=linux.LinuxUtilities, version=(2, 1, 0)
+            ),
+        ]
+
+    def _generator(
+        self,
+    ):
+        """Generates the list of block devices."""
+        vmlinux = self.context.modules[self.config["kernel"]]
+        try:
+            class_kset = vmlinux.object_from_symbol("class_kset")
+        except exceptions.SymbolError:
+            class_kset = None
+        if not class_kset:
+            raise TypeError(
+                "This plugin requires the class_kset structure. This structure is not present in the supplied symbol table. This means you are either analyzing an unsupported kernel version or that your symbol table is corrupt."
+            )
+
+        block_class = vmlinux.object_from_symbol("block_class")
+
+        for kobject in class_kset.list.to_list(
+            vmlinux.symbol_table_name + constants.BANG + "kobject", "entry"
+        ):
+            kset = linux.LinuxUtilities.container_of(kobject.vol.offset, "kset", "kobj", vmlinux)
+            subsys_private = linux.LinuxUtilities.container_of(kset.vol.offset, "subsys_private", "subsys", vmlinux)
+            if subsys_private.member("class") == block_class.vol.offset:
+                break
+
+        klist_devices = subsys_private.klist_devices
+        for device in self._device_iterator(self.context, klist_devices):
+            # Before v5.11, partitions are represented by hd_struct instead of block_device
+            if not vmlinux.has_type("hd_struct"):
+                block_device = linux.LinuxUtilities.container_of(device.vol.offset, "block_device", "bd_device", vmlinux)
+                gendisk = block_device.bd_disk
+
+                # lsblk by default skips over ram devices
+                try:
+                    if utility.array_to_string(gendisk.disk_name).startswith("ram"):
+                        continue
+                except exceptions.InvalidAddressException:
+                    continue
+            else:
+                hd_struct = linux.LinuxUtilities.container_of(device.vol.offset, "hd_struct", "__dev", vmlinux)
+                gendisk = linux.LinuxUtilities.container_of(hd_struct.vol.offset, "gendisk", "part0", vmlinux)
+                try:
+                    if utility.array_to_string(gendisk.disk_name).startswith("ram"):
+                        continue
+                except exceptions.InvalidAddressException:
+                    continue
+
+                block_device = self._get_block_device(vmlinux, device.devt)
+                if not block_device:
+                    continue
```

Comment:
It seems like you could refactor this into an internal helper function, such as `get_block_device()`, which would return either the `block_device` or `None`. This would help make the large for statement more readable and maintainable.

Reply:
hm sorry I just realized you already have a `_get_block_device()` .. so it's even more confusing ... you will need to think how to refactor these lines along with `_get_block_device()` to make it clearer.